### ISSUE TITLE
Handsontable cell is focusing, when focus in another input (for example, modal dialog)

### ISSUE THREAD

**Issue by dansabirov** [handsontable/handsontable#401]
Hi, @warpech !

Found interesting bug in new version of Handsontable.

I have a handsontable with `outsideClickDeselects: false` (I really need this option is set to false), and jquery ui dialog. When I select any cell in handsontable and showing my dialog (by pressing button) and trying to press in my input on dialog by mouse focus is setting to selected cell in handsontalbe.

Here is demonstration code http://jsfiddle.net/ggwgk/8/

Btw, there is no such problem in version 0.7.5

**Comment by floribon** [comment]
Same issue here, I need to programatically set the focus on another `input` but the first key I press write inside the selected cell.

Unfortunately handsontable is part of our build process so I cannot edit it, unless I fork which I would love to avoid for such a minor thing.

**Comment by AMBudnik** [comment]
Hi guys, please update to 0.20.0 if possible, before that version we had some issues with [focus](https://github.com/handsontable/handsontable/issues/2961) it can be related to what you face with.

**Comment by floribon** [comment]
Hi @AMBudnik I've upgraded to 0.20.0, unfortunately I still have that problem. I'm using jQUery to focus an input `$('#myinput').focus()` and I see the text blinking cursor on it, but when I type something the text get entered in the cell instead.

If I may, I guess you have a keypress listener, on which you could check if the focus is on the current cell before adding text to it.

**Comment by alexisszabo** [comment]
I think @AMBudnik meant to try to 0.20.1

**Comment by AMBudnik** [comment]
Thank you @alexisszabo yes I meant 0.20.1 where on of focus bugs were fixed.

**Comment by kemins** [comment]
The same issue with 0.20.1.

**Comment by floribon** [comment]
Indeed, I forgot to reply but I actually upgraded to the latest at that moment, 0.20.1, which didn't solve the issue.

**Comment by AMBudnik** [comment]
Could you create a simple demo showing this bug with latest version applied?

**Comment by floribon** [comment]
Here it is:

http://plnkr.co/edit/7jpkMMoHtUT66SlL99F3?p=preview

I believe we could avoid it by "unselecting" the cell within the table first, if the API allows that.

But at best I would need the cell to keep being selected, with the blue border, yet still type text where the focus is.

**Comment by AMBudnik** [comment]
present in 0.24.1

**Comment by floribon** [comment]
Still present in `0.26.1`

As a reminder, here is how to reproduce: http://plnkr.co/edit/7jpkMMoHtUT66SlL99F3?p=preview

**Comment by AMBudnik** [comment]
thanks for feedback @floribon

**Comment by swistach** [comment]
Hi @floribon 
We have to discuss about this issue in the core-team, but for now we can recommend a workaround:  http://plnkr.co/edit/lHsfQwhHLgJgdLOW1ntN?p=preview

**Comment by floribon** [comment]
Hi @swistach, clear and simple, thanks!

**Comment by zoladkow** [comment]
Hi,
Another curious variation on this subject: when the table is set to readOnly, input outside the table gets some of the keystrokes - letters etc. But arrows, delete & enter are still only captured on the table.

That is my use case - a readonly table, with a filter sidenav which should get the focus/input once the sidenav expands. Table selection should stay as indication where we are, on what row some action will perform - hence outsideClickDeslects: false. At least let the currentRowClassName stay on...

Modified workaround from swistach which shows this:
http://plnkr.co/edit/UKEhnk9O8FwOWD9J5x8u?p=preview

Oh, I bumped the example to 0.31.0

**Comment by swistach** [comment]
Hi @zoladkow,
in your case, you have to use `unlisten()`: http://plnkr.co/edit/Jn9gHBgCTUjEkuhb9157?p=preview

Selection will stay without any changes but all mounted events will by stopped. When you click on the table, you will set again listening on the Handsontable.

Additionally you can remove `outsideClickDeselect` and then you can write something without losing cell selection, but when someone click outside of the table, then selection will be removed.

**Comment by pmunin** [comment]
How come it's closed if the bug still exists?
Is there any workaround for situation when the trigger of focus does not know anything about HoT instance? There is a plugin on my page that listens globally the Ctrl+F event and sets focus to its input. Can I somehow subscribe to blur event of Hot and invoke "deselectCell" there, so it will work for every invokers of otherControl.focus()?

**Comment by AMBudnik** [comment]
Hi @pmunin 

Above @swistach suggested [`unlisten`](https://docs.handsontable.com/pro/1.18.0/Core.html#unlisten) method. Have you tried it?

**Comment by pmunin** [comment]
@AMBudnik "unlisten" is a method of handsontable instance, which CTRL+F overrider have no access to - it's different 3rd party module. Is there any way to catch event of Handsontable loosing focus to execute "unlisten" there?

**Comment by AMBudnik** [comment]
That is a good question. Unfortunately, I don't know an answer for it but will ask our developer for some hints.

**Comment by AMBudnik** [comment]
Hi @pmunin 

It looks like we would need to put some more logic.

Here I have a demo that works as expected http://plnkr.co/edit/CajKBFzyIAlkphVtDuep?p=preview but... it adds the `editor: false` (line 36) which disables the cell edition.
What you can test is adding the `editor: true` when a user uses <kbd>Enter</kbd>. I am not sure if that meets your project requirements but that's an option that should work.

**Comment by pmunin** [comment]
@AMBudnik your code does not work, sorry - HoT still intercepts keys like: Arrow(Left/Right/Up/Down), Delete, Backspace, CTRL+A (Select all) after moving focus to Input via CTRL+F

**Comment by AMBudnik** [comment]
Maybe I'm doing something wrong. 

Here is this demo on Chrome/Windows 10 doing the steps listed there

![gif3](https://user-images.githubusercontent.com/10757813/37714804-bbfc6056-2d1a-11e8-8ce4-b3aad2161216.gif)

**Comment by pmunin** [comment]
1) Select a cell in the HoT
2) Press CTRL+F (Focus moves to Input box)
3) In Input box enter "hallo world". Press Backspace
Expected results: delete last letter "d" from the input box
Actual results: Removes HoT cell's content

**Comment by AMBudnik** [comment]
So that was what I've missed. You're right. I've used <kbd>Backspace</kbd> and it erased my cell instead of the input.
I am marking this topic as a *bug*

**Comment by kevincolyar** [comment]
Not a fix but this was my workaround:

```javascript
outsideClickDeselects: (event) => {
  // Deselect cells in grid unless the clicked element matches one of the listed selectors
  // e.g. Grid control buttons or the dialog/popover they open
  return !event.closest('.grid button, .bp3-dialog, .bp3-popover')
},
```

It let's you specify which clicked elements should or shouldn't deselect cells in the grid.

**Comment by AMBudnik** [comment]
Thank you for sharing the workaround @kevincolyar

**Comment by GaelPaquet** [comment]
Hi, any update on this very annoying bug ?

**Comment by AMBudnik** [comment]
You're right @GaelPaquet the issue had its ups and downs but it is old and still present. We had some attempts on fixing it (`in development` label on 21th of March) but it failed. I will discuss this subject with our CTO and share an update.

**Comment by AMBudnik** [comment]
The issue for `<input>` should be fixed with https://github.com/handsontable/handsontable/pull/5631 (in the next release. The `contenteditable="true"` is another that that is not fixed within that snippet. That is why I am opening another issue for that bug.

**Comment by AMBudnik** [comment]
It looks like the modal window + `contenteditable="true"` also does not allow us to replicate the issue, so the new issue was not created. I will close this one only after we publish the new version and check it once more.

**Comment by AMBudnik** [comment]
Hey guys!
@dansabirov @bruceh48 @GaelPaquet @kevincolyar @pmunin @floribon @kemins @alexgann @Cirunz @codename- @dansabirov @brucelandgate 

We've just fixed a very similar issue here https://github.com/handsontable/handsontable/issues/5541 for the Handsontable  `7.0.0`. Can you confirm that it works for you?

**Comment by AMBudnik** [comment]
I believe that we can close the topic as there was no reply for more than 2 weeks, but if something is still wrong please let me know.

**Comment by kevincolyar** [comment]
I'm still having this issue on v7.0.0

**Comment by AMBudnik** [comment]
Thank you for a reply, @kevincolyar 
I want to narrow down the issue as much as possible. 

Does your workaround

```js
outsideClickDeselects: (event) => {
  // Deselect cells in grid unless the clicked element matches one of the listed selectors
  // e.g. Grid control buttons or the dialog/popover they open
  return !event.closest('.grid button, .bp3-dialog, .bp3-popover')
},
```

have any impact on the 7.0.0 or it works the same (not better/not worse)?

I am reopening the topic.

**Comment by kevincolyar** [comment]
My workaround doesn't work in the case where:
1. A handsontable cell is selected.
2. A button is clicked to open a dialog **with an text input automatically in focus**.
3. Any input intended for that dialog input box will be sent to the selected handsontable cell behind the dialog

![image](https://user-images.githubusercontent.com/17285/54834472-7f303700-4c7d-11e9-85f9-4c0e572b05d7.png)

**Comment by kevincolyar** [comment]
I believe I've found a work-around that solves this issue.

Adding the following key handler to the Handsontable constructor will stop it from stealing input from the specified container elements:

```javascript

beforeKeyDown: (event) => {
  // Prevent Handsontable from stealing input from the following containers
  if(event.target.closest('.controls, .bp3-dialog, .bp3-popover, .bp3-overlay-backdrop, .bp3-fill'))
    event.stopImmediatePropagation()
},
```

**Comment by ottosson** [comment]
I have this issue as well. Using the mouse to select an input on a modal will work for me, but not setting the focus programmatically...

**Comment by mewalig** [comment]
worked for me, maybe works w locked cells? Thanks @kevincolyar and @ottosson for the tips
```
beforeKeyDown: function(e) => {
   if (!e.target.closest(".handsontable")) {
      e.stopImmediatePropagation();
      return;
   }
}
```

**Comment by AMBudnik** [comment]
It works the same in v 7.4.2 https://jsfiddle.net/sqou6m7p/1/ and 8.0.0-beta2-rev18 https://jsfiddle.net/sqou6m7p/2/

**Comment by hernanfaggiani** [comment]
Hello. Is there any solution for this issue. None of the proposed solutions worked for me.
 * Version: 8.4.0
 * Release date: 11/05/2021 (built at 11/05/2021 12:29:54)

**Comment by AMBudnik** [comment]
Hi @hernanfaggiani this issue hasn't been officially solved yet.
I will make sure to notify you once we fix it.

**Comment by AMBudnik** [comment]
This issue does not seem to be replicable anymore. Could I ask any of you

@dansabirov @hernanfaggiani @ottosson @GaelPaquet  
@mewalig @kevincolyar @pmunin
@bruceh48 @floribon @kemins @alexgann @Cirunz @codename- @dansabirov @brucelandgate

to confirm? If it still remains for you, please let me know and we will reopen the issue report.

**Comment by mohan-provus** [comment]
This issue still there in version ## [12.2.0] - 2022-10-25

**Comment by ottosson** [comment]
After just a quick check it seems to be fixed. Thanks!

**Comment by AMBudnik** [comment]
@ottosson I am more than happy to hear that!

@mohan-provus Could you send more details about the use case to support@handsontable.com? Including device type and version?

**Comment by gdiazcamilo** [comment]
This issue is present in the following scenario that @kevincolyar was describing in version 14.5:

1. A handsontable cell is selected.
2. A button is clicked to open a dialog with a text input automatically in focus.
3. Space or backspace keys are sent to the handsontable cell and not to the input with the focus in the modal dialog.

My workaround was to call `hot.deselectCell` after the modal opened.

**Comment by adrianszymanski89** [comment]
Hi @gdiazcamilo 

Thank you for contacting us. This issue has already been solved. If you are still encountering it, please share a code demo where it can be replicated.

**Comment by gdiazcamilo** [comment]
Here's a code demo where you can see the error: https://jsfiddle.net/n1hqe8x3/
Steps:
1. Click any cell in the table
2. Click the "Open modal" button. After the modal is open the focus will be applied to the input text.
3. Start typing. You can type letters or numbers, but if you type a space or press the backspace key the space will be typed in the handsontable selected cell.

If you click the input, it works fine, the problem is when you focus the input programmatically.

**Comment by ottosson** [comment]
> This issue has already been solved. If you are still encountering it, please share a code demo where it can be replicated.

@adrianszymanski89 I still experience this issue as well.

**Comment by adrianszymanski89** [comment]
Hi @gdiazcamilo @ottosson 

Thank you for the update. I can confirm that the issue is still present. We have it reported internally, and it awaits the fix. I'll update you as soon as it's done.

**Comment by d-haber** [comment]
Hi @adrianszymanski89.  We have had to apply a workaround to deal with this. Do you have any idea where on the roadmap this issue falls?

Thanks.

**Comment by sl01k** [comment]
This bug appears to be fixed in v17.0.0

Test result: All keystrokes (including Space and Backspace) were correctly received by the modal input. The Handsontable cell remained empty and the editor was not activated.
